### ISSUE TITLE
 cannot load library 'pango-1.0': error 0x7e. Additionally, ctypes.util.find_library() did not manage to locate a library called 'pango-1.0'

### ISSUE THREAD

**Issue by nithinreddyy** [Kozea/WeasyPrint#1565]
I installed the Weasyprint library. 

When i ran 'python -m pip install weasyprint', it got installed. Below is the image for your reference

<img width="615" alt="1" src="https://user-images.githubusercontent.com/56256685/153348282-11804359-b59b-4116-8906-1cc36c79a37d.PNG">
<img width="151" alt="2" src="https://user-images.githubusercontent.com/56256685/153348293-cfa97bae-e399-4b5e-ba55-b882bd86f250.PNG">

When i'm running 'python -m weasyprint --info' , it's returning an error present in the below image
<img width="665" alt="3" src="https://user-images.githubusercontent.com/56256685/153348649-1c0ab3a6-d2ce-4c79-81f5-01f1f5841cd4.PNG">

I'm using windows 10, anaconda, Python 3.7.x version with GPU enabled

**How to overcome this issue?**

**Comment by liZe** [comment]
Hello!

Could you please try the latest version of WeasyPrint (54.1) with a more recent version of Python (at least 3.8)? A lot of things have been improved about Windows support with these versions.

**Comment by nithinreddyy** [comment]
> Hello!
> 
> Could you please try the latest version of WeasyPrint (54.1) with a more recent version of Python (at least 3.8)? A lot of things have been improved about Windows support with these versions.

I tried with python 3.8.5 version and installed weasyprint 54.1 version. But it's still throwing the same error. The error is below

```
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
Input In [1], in <module>
----> 1 from doctr.models import ocr_predictor

File ~\miniconda3\envs\weasy\lib\site-packages\doctr\__init__.py:1, in <module>
----> 1 from . import datasets, io, models, transforms, utils
      2 from .file_utils import is_tf_available, is_torch_available
      3 from .version import __version__

File ~\miniconda3\envs\weasy\lib\site-packages\doctr\datasets\__init__.py:3, in <module>
      1 from doctr.file_utils import is_tf_available
----> 3 from .generator import *
      4 from .cord import *
      5 from .detection import *

File ~\miniconda3\envs\weasy\lib\site-packages\doctr\datasets\generator\__init__.py:4, in <module>
      1 from doctr.file_utils import is_tf_available, is_torch_available
      3 if is_tf_available():
----> 4     from .tensorflow import *
      5 elif is_torch_available():
      6     from .pytorch import *

File ~\miniconda3\envs\weasy\lib\site-packages\doctr\datasets\generator\tensorflow.py:8, in <module>
      1 # Copyright (C) 2021, Mindee.
      2 
      3 # This program is licensed under the Apache License version 2.
      4 # See LICENSE or go to <https://www.apache.org/licenses/LICENSE-2.0.txt> for full license details.
      6 import tensorflow as tf
----> 8 from .base import _CharacterGenerator, _WordGenerator
     10 __all__ = ['CharacterGenerator', 'WordGenerator']
     13 class CharacterGenerator(_CharacterGenerator):

File ~\miniconda3\envs\weasy\lib\site-packages\doctr\datasets\generator\base.py:11, in <module>
      7 from typing import Any, Callable, List, Optional, Tuple, Union
      9 from PIL import Image, ImageDraw
---> 11 from doctr.io.image import tensor_from_pil
     12 from doctr.utils.fonts import get_font
     14 from ..datasets import AbstractDataset

File ~\miniconda3\envs\weasy\lib\site-packages\doctr\io\__init__.py:2, in <module>
      1 from .elements import *
----> 2 from .html import *
      3 from .image import *
      4 from .pdf import *

File ~\miniconda3\envs\weasy\lib\site-packages\doctr\io\html.py:8, in <module>
      1 # Copyright (C) 2021, Mindee.
      2 
      3 # This program is licensed under the Apache License version 2.
      4 # See LICENSE or go to <https://www.apache.org/licenses/LICENSE-2.0.txt> for full license details.
      6 from typing import Any
----> 8 from weasyprint import HTML
     10 __all__ = ['read_html']
     13 def read_html(url: str, **kwargs: Any) -> bytes:

File ~\miniconda3\envs\weasy\lib\site-packages\weasyprint\__init__.py:325, in <module>
    322         yield 'string', string, base_url, None
    324 # Work around circular imports.
--> 325 from .css import preprocess_stylesheet  # noqa isort:skip
    326 from .html import (  # noqa isort:skip
    327     HTML5_UA_COUNTER_STYLE, HTML5_UA_STYLESHEET, HTML5_PH_STYLESHEET,
    328     find_base_url)
    329 from .document import Document, Page

File ~\miniconda3\envs\weasy\lib\site-packages\weasyprint\css\__init__.py:27, in <module>
     25 from ..logger import LOGGER, PROGRESS_LOGGER
     26 from ..urls import URLFetchingError, get_url_attribute, url_join
---> 27 from . import computed_values, counters, media_queries
     28 from .properties import INHERITED, INITIAL_NOT_COMPUTED, INITIAL_VALUES
     29 from .utils import get_url, remove_whitespace

File ~\miniconda3\envs\weasy\lib\site-packages\weasyprint\css\computed_values.py:16, in <module>
     13 from tinycss2.color3 import parse_color
     15 from ..logger import LOGGER
---> 16 from ..text.ffi import ffi, pango, units_to_double
     17 from ..text.line_break import Layout, first_line_metrics, line_size
     18 from ..urls import get_link_attribute

File ~\miniconda3\envs\weasy\lib\site-packages\weasyprint\text\ffi.py:406, in <module>
    401             pass
    403 gobject = _dlopen(
    404     ffi, 'gobject-2.0-0', 'gobject-2.0', 'libgobject-2.0-0',
    405     'libgobject-2.0.so.0', 'libgobject-2.0.dylib',  'libgobject-2.0-0.dll')
--> 406 pango = _dlopen(
    407     ffi, 'pango-1.0-0', 'pango-1.0', 'libpango-1.0-0', 'libpango-1.0.so.0',
    408     'libpango-1.0.dylib', 'libpango-1.0-0.dll')
    409 harfbuzz = _dlopen(
    410     ffi, 'harfbuzz', 'harfbuzz-0.0', 'libharfbuzz-0',
    411     'libharfbuzz.so.0', 'libharfbuzz.so.0', 'libharfbuzz.0.dylib',
    412     'libharfbuzz-0.dll')
    413 fontconfig = _dlopen(
    414     ffi, 'fontconfig-1', 'fontconfig', 'libfontconfig', 'libfontconfig.so.1',
    415     'libfontconfig-1.dylib', 'libfontconfig-1.dll')

File ~\miniconda3\envs\weasy\lib\site-packages\weasyprint\text\ffi.py:390, in _dlopen(ffi, *names)
    388         pass
    389 # Re-raise the exception.
--> 390 return ffi.dlopen(names[0])

File ~\AppData\Roaming\Python\Python38\site-packages\cffi\api.py:150, in FFI.dlopen(self, name, flags)
    147     raise TypeError("dlopen(name): name must be a file name, None, "
    148                     "or an already-opened 'void *' handle")
    149 with self._lock:
--> 150     lib, function_cache = _make_ffi_library(self, name, flags)
    151     self._function_caches.append(function_cache)
    152     self._libraries.append(lib)

File ~\AppData\Roaming\Python\Python38\site-packages\cffi\api.py:832, in _make_ffi_library(ffi, libname, flags)
    830 def _make_ffi_library(ffi, libname, flags):
    831     backend = ffi._backend
--> 832     backendlib = _load_backend_lib(backend, libname, flags)
    833     #
    834     def accessor_function(name):

File ~\AppData\Roaming\Python\Python38\site-packages\cffi\api.py:827, in _load_backend_lib(backend, name, flags)
    825     if first_error is not None:
    826         msg = "%s.  Additionally, %s" % (first_error, msg)
--> 827     raise OSError(msg)
    828 return backend.load_library(path, flags)

OSError: cannot load library 'pango-1.0-0': error 0x7e.  Additionally, ctypes.util.find_library() did not manage to locate a library called 'pango-1.0-0'

```

<img width="781" alt="Capture" src="https://user-images.githubusercontent.com/56256685/153476817-40f341da-4731-4140-92f6-c4daa49475d2.PNG">

**Comment by liZe** [comment]
You’re using Miniconda. It may work, but it’s not supported and we’ve never tried to make WeasyPrint work with it.

If you want to install WeasyPrint on Windows, please follow the [installation steps of the documentation](https://doc.courtbouillon.org/weasyprint/stable/first_steps.html#windows). If you want to follow a different path, it’s OK but then you’re on your own 😒.

**Comment by nithinreddyy** [comment]
> 

I tried with normal python with command prompt with 3.8.5 version, installed weasyprint. When i'm running '**python -m weasyprint --info**', it's returning an error. Below is the image for your reference

<img width="695" alt="Capture" src="https://user-images.githubusercontent.com/56256685/153555495-0d0469b5-e642-43b2-a733-3331e9d3b48d.PNG">

**Comment by liZe** [comment]
Your error message is about msys64, that’s unrelated to WeasyPrint or its dependencies. There’s probably an incompatibility between the libraries included in msys64 and the ones in GTK.

We’re really sorry, but there’s no way for us to address all the problems WeasyPrint can have when other conflicting tools are also installed on the same computer. Our install instructions work on fresh installs of Windows, and managing tools installed in parallel is beyond our knowledge and our capacities.

**Comment by fg-mindee** [comment]
Hi @liZe :wave: 

Thanks a lot for this great project :pray: 

To add some context, the original issue of @nithinreddyy was this one https://github.com/mindee/doctr/issues/815
After checking a few things, I suggested checking on weasyprint side since the traceback was pointing to it.

Anyway, I'd love to help so what would be your suggestions to solve this? I understand you're suggesting it's a dependency conflict specific to Windows for the last installation step you suggested. To the best of my knowledge, @nithinreddyy tried to use the Windows-specific installation instructions in your documentation.

As an open-source developer, I completely understand that there are always grey areas for some users' setup. I'm only trying to identify:
- whether the installation procedure is correct for Windows
- if so, are there some precautions that need to be taken to avoid the installation procedure to fail (is `ligbmodule` a normal dependency for weasyprint on Windows?)
- if not, what are our leads to address the `pango` error?

If needed we can always try to reproduce this in a docker environment :man_shrugging:

**Comment by liZe** [comment]
> Hi @liZe wave

Hi!

> Anyway, I'd love to help so what would be your suggestions to solve this? I understand you're suggesting it's a dependency conflict specific to Windows for the last installation step you suggested. To the best of my knowledge, @nithinreddyy tried to use the Windows-specific installation instructions in your documentation.

The error message shows that @nithinreddyy used msys64’s version of Python. That’s not what’s in the [installation guide for Windows](https://doc.courtbouillon.org/weasyprint/latest/first_steps.html#windows). I don’t know what’s "wrong" in msys64’s `libgmodule-2.0-0.dll`, but I know that it works when you use Microsoft Store’s Python and GTK.

If you really want to use msys64, you can take a look at the [CI](https://github.com/Kozea/WeasyPrint/blob/master/.github/workflows/tests.yml), that actually uses msys64. But that’s not a supported way of using WeasyPrint, so if it doesn’t work you’re on your own :/.

> * whether the installation procedure is correct for Windows

Following the installation guide is the best way to get WeasyPrint working on Windows.

> * if so, are there some precautions that need to be taken to avoid the installation procedure to fail (is `ligbmodule` a normal dependency for weasyprint on Windows?)

libgmodule is not a direct dependency of WeasyPrint. It may be a dependency of a dependency.

> * if not, what are our leads to address the `pango` error?

Don’t use Miniconda (as in the original comment), don’t use msys64 (as in the latest comment). Just follow the installation guide. It doesn’t mean that WeasyPrint can’t work using Conda or msys64, but it means that if you want to follow different installation paths I won’t be able to help you.

**Comment by fg-mindee** [comment]
Thanks for the explanations @liZe :pray: 
@nithinreddyy have you tried to install this in a clean environment? (apparently not a conda environment)

**Comment by morlandi** [comment]
Same problem on Ubuntu 20.
Solved with this:

```
apt-get install -y libpangocairo-1.0-0
```

**Comment by liZe** [comment]
> Same problem on Ubuntu 20. Solved with this:
> 
> ```
> apt-get install -y libpangocairo-1.0-0
> ```

On Ubuntu, following the installation guide works. The problem is different here, as it’s related to Windows.

@fg-mindee @nithinreddyy I’m closing this issue, as using a clean Windows environment should work. Feel free to add new comments if there’s something more we can do for you.

**Comment by unlessbamboo** [comment]
> 
Thanks.
system: MacOS M1
docker system: ubuntu 20.04
Action: Solve this proble after upgrade weasyprint from 52 to 53.4.

**Comment by dimanskr** [comment]
I made a django website for windows 10 with the weasyprint library. Also installed GTK+ for Windows Runtime Environment (version 3). The same errors were thrown.
I tried to swap paths in path, updated windows to 11, but it didn't help.
Then i pasted 'sitecustomize.py' like **Tontyna** https://github.com/Kozea/WeasyPrint/issues/971#issuecomment-544195744 to the virtual environment folder '/Lib/site-packages'
another error occurred: Fontconfig error: Cannot load default config file.
It turned out that the 'C:\Users\<User>\anaconda3\fontconfig.dll' library is being loaded.
In the file 'sitecustomize.py' I changed the line:
`os.environ['PATH'] = GTK_FOLDER + os.pathsep + os.environ.get('PATH', '') `
to the line: 
`os.environ['PATH'] = GTK_FOLDER`
and it WORKS

**Comment by mohammadanarul** [comment]
@nithinreddyy
This problem: OSError: cannot load library 'pango-1.0-0': pango-1.0-0: cannot open shared object file: No such file or directory.  Additionally, ctypes.util.find_library() did not manage to locate a library called 'pango-1.0-0

## `20.04.2-Ubuntu` Solution below 
`sudo apt-get install -y libpangocairo-1.0-0`

**Comment by kopyl** [comment]
@mohammadanarul thank you very much, this ^ helped

**Comment by liZe** [comment]
If you have this problem, the best solution is to read the documentation, with a [complete installation guide](https://doc.courtbouillon.org/weasyprint/stable/first_steps.html#installation) that includes the dependences that have to be installed. Please read the documentation before adding a comment here 😄, and open a new issue if there’s something wrong with the documentation.

**Comment by wabiloo** [comment]
I have the same problem on MacOS, after doing `brew install weasyprint`, which installed version 61.2

**Comment by liZe** [comment]
> I have the same problem on MacOS, after doing `brew install weasyprint`, which installed version 61.2

Then it’s a problem with your Homebrew installation, and there’s nothing we can do for you here. A fresh installation of WeasyPrint on Homebrew works.

If you have this problem, the best solution is to read the documentation, with a [complete installation guide](https://doc.courtbouillon.org/weasyprint/stable/first_steps.html#installation) that includes the dependences that have to be installed. Please read the documentation and open a new issue if there’s something wrong with the documentation.
